src/main/java/net/neoforged/neoforge/event/BuildCreativeModeTabContentsEvent.java

```diff
@@ -61,12 +64,127 @@ public boolean hasPermissions() {
+     */
+    public List<ItemStack> getSearchEntries() {
+        return this.searchEntries.stream().toList();
     }
 
+    /**
+     * Inserts the new stack at the end of the given tab at this point in time.
+     * 
+     * @exception IllegalArgumentException if the new itemstack's count is not 1.
+     */
     @Override
-    public void accept(ItemStack stack, CreativeModeTab.TabVisibility visibility) {
-        getEntries().put(stack, visibility);
+    public void accept(ItemStack newEntry, CreativeModeTab.TabVisibility visibility) {
+        assertStackCount(newEntry);
+
+        if (isParentTab(visibility)) {
+            parentEntries.add(newEntry);
+        }
+
+        if (isSearchTab(visibility)) {
+            searchEntries.add(newEntry);
+        }
+    }
+
+    /**
+     * Inserts the new entry after the specified existing entry.
+     * 
+     * @exception UnsupportedOperationException if the existing entry is not found in the tab's lists.
+     * @exception IllegalArgumentException      if the new itemstack's count is not 1 or target does not exist in set.
+     */
+    public void putAfter(ItemStack existingEntry, ItemStack newEntry, CreativeModeTab.TabVisibility visibility) {
+        assertStackCount(newEntry);
+
+        if (isParentTab(visibility)) {
+            assertTargetExists(parentEntries, existingEntry);
+            parentEntries.addBefore(existingEntry, newEntry);
+        }
+
+        if (isSearchTab(visibility)) {
+            assertTargetExists(searchEntries, existingEntry);
+            searchEntries.addBefore(existingEntry, newEntry);
+        }
+    }
+
+    /**
+     * Inserts the new entry before the specified existing entry.
+     * 
+     * @exception UnsupportedOperationException if the existing entry is not found in the tab's lists.
+     * @exception IllegalArgumentException      if the new itemstack's count is not 1 or target does not exist in set.
+     */
+    public void putBefore(ItemStack existingEntry, ItemStack newEntry, CreativeModeTab.TabVisibility visibility) {
+        assertStackCount(newEntry);
+
+        if (isParentTab(visibility)) {
+            assertTargetExists(parentEntries, existingEntry);
+            parentEntries.addAfter(existingEntry, newEntry);
+        }
+
+        if (isSearchTab(visibility)) {
+            assertTargetExists(searchEntries, existingEntry);
+            searchEntries.addAfter(existingEntry, newEntry);
+        }
+    }
+
+    /**
+     * Inserts the new entry in the front of the tab's content.
+     * 
+     * @exception IllegalArgumentException if the new itemstack's count is not 1.
+     */
+    public void putFirst(ItemStack newEntry, CreativeModeTab.TabVisibility visibility) {
+        assertStackCount(newEntry);
+
+        if (isParentTab(visibility)) {
+            parentEntries.addFirst(newEntry);
+        }
+
+        if (isSearchTab(visibility)) {
+            searchEntries.addFirst(newEntry);
+        }
```

Comment:
If a vanilla item is input that already exists in this tab, then it will be moved to the front of the tab

Reply:
I'll add duplicate validation checks. In theory, devs should remove the vanilla entry first. And then insert it where they want instead of relying on hidden side effects.